lib/FederationService.php

```diff
@@ -0,0 +1,215 @@
+			'&accessToken=' . $accessToken;
+		return $remoteFileUrl;
+	}
+	
+	/**
+	*
+	* @param string $server addres of a remote server
+	* @param string $accessToken wopi access token from a remote server
+	* @return array|null with additional wopi information
+	*/
+	public function getWopiForToken($server, $accessToken) {
+		$remote = $server;
+
+		if (!$this->isServerAllowed($remote)) {
+			$this->logger->info("Server {server} is not allowed.", ["server" => $remote]);
+			return null;
+		}
+
+		try {
+			$client = $this->httpClient->newClient();
+			$url = $remote . '/ocs/v2.php/apps/richdocuments/api/v1/federation';
+
+			$response = $client->post(
+				$url,
+				[
+					'form_params' => [
+						'token' => $accessToken,
+						'format' => 'json'
+					],
+					'timeout' => 3,
+					'connect_timeout' => 3,
+				]
+			);
+
+			$responseBody = $response->getBody();
+			$data = \json_decode($responseBody, true, 512);
+			if (\is_array($data)) {
+				return $data['ocs']['data'];
+			}
+			return null;
+		} catch (\Throwable $e) {
+			$this->logger->info('Cannot get the wopi info from remote server: ' . $remote, ['exception' => $e]);
+		}
+
+		return null;
+	}
+
+	/**
+	 * Check if server is allowed
+	 *
+	 * @param string $remote a remote url
+	 * @return bool indicating if given remote is allowed server
+	 */
+	public function isServerAllowed($remote) {
+		// TODO: implement check for trusted server, for a moment all trusted
+
+		return true;
+	}
+
+	/**
+	 * Get the wopiSrc Url from a remote server.
+	 *
+	 * @param string $server a remote
+	 * @return string with the wopi src Url
+	 */
+	public function getRemoteWopiSrc($server) {
+		if (\strpos($server, 'http://') === false && \strpos($server, 'https://') === false) {
+			$server = 'https://' . $server;
+		}
+
+		if (!$this->isServerAllowed($server)) {
+			$this->logger->info("Server {server} is not allowed.", ["server" => $server]);
+			return '';
+		}
+
+		try {
+			$getWopiSrcUrl = $server . '/ocs/v2.php/apps/richdocuments/api/v1/federation?format=json';
+			$client = $this->httpClient->newClient();
+			$response = $client->get($getWopiSrcUrl, ['timeout' => 5]);
+			$data = \json_decode($response->getBody(), true);
+
+			if (\is_array($data)) {
+				return $data['ocs']['data']['wopi_url'];
+			}
+		} catch (\Throwable $e) {
+			$this->logger->info('Cannot get the wopiSrc of remote server: ' . $server, ['exception' => $e]);
+		}
+
+		return '';
+	}
+
+	/**
+	 * split user and remote from federated cloud id, null if not federated cloud id
+	 *
+	 * @param string $userId user id
+	 * @return string
+	 */
+	public function generateFederatedCloudID(string $userId) : string {
+		if (\strpos($userId, '@') === false) {
+			// generate federated cloud id
+			$user =  $userId;
+			$remote = \preg_replace('|^(.*?://)|', '', \rtrim($this->urlGenerator->getAbsoluteURL('/'), '/'));
+			return "{$user}@{$remote}";
+		}
+
+		// Find the first character that is not allowed in user names
```

Comment:
Too many little things here... maybe move this part to a specialized method so we can add unit tests?
In addition, smaller methods are easier to explain what they do.

Reply:
it is already tested in different class/app. I could potentially used federation app class here with all these reflection stuff.. see above.